File: backend/task_management.py
"""
TASK MANAGEMENT SYSTEM
Advanced task scheduling and execution for Buffr Host operations
"""

from typing import Dict, List, Optional, Any, Union, Callable
from datetime import datetime, timedelta
from enum import Enum
from dataclasses import dataclass, field
from sqlalchemy import Column, Integer, String, DateTime, JSON, Boolean, ForeignKey, Text, Float
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship
import asyncio
import json
import uuid
import threading
from concurrent.futures import ThreadPoolExecutor

Base = declarative_base()

class TaskStatus(Enum):
    """Task execution status"""
    PENDING = "pending"
    SCHEDULED = "scheduled"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    RETRYING = "retrying"

class TaskPriority(Enum):
    """Task priority levels"""
    LOW = 1
    NORMAL = 2
    HIGH = 3
    URGENT = 4
    CRITICAL = 5

class TaskType(Enum):
    """Task types"""
    EMAIL = "email"
    NOTIFICATION = "notification"
    REPORT = "report"
    CLEANUP = "cleanup"
    BACKUP = "backup"
    SYNC = "sync"
    ANALYSIS = "analysis"
    CUSTOM = "custom"

@dataclass
class TaskConfig:
    """Task configuration"""
    max_retries: int = 3
    retry_delay: int = 60  # seconds
    timeout: int = 300  # seconds
    priority: TaskPriority = TaskPriority.NORMAL
    dependencies: List[str] = field(default_factory=list)
    tags: List[str] = field(default_factory=list)
    metadata: Dict[str, Any] = field(default_factory=dict)

class Task(Base):
    """Task model"""
    __tablename__ = 'tasks'
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(255), nullable=False)
    description = Column(Text)
    type = Column(String(50), nullable=False)
    status = Column(String(50), default=TaskStatus.PENDING.value)
    priority = Column(Integer, default=TaskPriority.NORMAL.value)
    config = Column(JSON)  # TaskConfig object
    payload = Column(JSON)  # Task data
    scheduled_at = Column(DateTime)
    started_at = Column(DateTime)
    completed_at = Column(DateTime)
    retry_count = Column(Integer, default=0)
    max_retries = Column(Integer, default=3)
    error_message = Column(Text)
    result = Column(JSON)
    created_by = Column(String(255))
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    dependencies = relationship("TaskDependency", back_populates="task")

class TaskDependency(Base):
    """Task dependency model"""
    __tablename__ = 'task_dependencies'
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    task_id = Column(String, ForeignKey('tasks.id'))
    depends_on_task_id = Column(String, ForeignKey('tasks.id'))
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    task = relationship("Task", back_populates="dependencies")

class TaskExecution(Base):
    """Task execution log model"""
    __tablename__ = 'task_executions'
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    task_id = Column(String, ForeignKey('tasks.id'))
    status = Column(String(50), nullable=False)
    started_at = Column(DateTime, default=datetime.utcnow)
    completed_at = Column(DateTime)
    duration = Column(Float)  # seconds
    error_message = Column(Text)
    result = Column(JSON)
    worker_id = Column(String(255))

class TaskManager:
    """Advanced task management system"""
    
    def __init__(self, db_session, max_workers: int = 10):
        self.db = db_session
        self.max_workers = max_workers
        self.executor = ThreadPoolExecutor(max_workers=max_workers)
        self.task_handlers: Dict[str, Callable] = {}
        self.running_tasks: Dict[str, asyncio.Task] = {}
        self.scheduler_running = False
        self._register_default_handlers()
    
    def _register_default_handlers(self):
        """Register default task handlers"""
        self.task_handlers.update({
            'email': self._handle_email_task,
            'notification': self._handle_notification_task,
            'report': self._handle_report_task,
            'cleanup': self._handle_cleanup_task,
            'backup': self._handle_backup_task,
            'sync': self._handle_sync_task,
            'analysis': self._handle_analysis_task,
            'custom': self._handle_custom_task
        })
    
    async def create_task(self, task_data: Dict[str, Any]) -> Task:
        """Create a new task"""
        try:
            config = TaskConfig(**task_data.get('config', {}))
            
            task = Task(
                name=task_data['name'],
                description=task_data.get('description', ''),
                type=task_data['type'],
                priority=config.priority.value,
                config=config.__dict__,
                payload=task_data.get('payload', {}),
                scheduled_at=task_data.get('scheduled_at'),
                max_retries=config.max_retries,
                created_by=task_data.get('created_by', 'system')
            )
            
            self.db.add(task)
            await self.db.commit()
            await self.db.refresh(task)
            
            # Create dependencies
            if config.dependencies:
                await self._create_dependencies(task.id, config.dependencies)
            
            return task
        except Exception as e:
            await self.db.rollback()
            raise Exception(f"Failed to create task: {str(e)}")
    
    async def _create_dependencies(self, task_id: str, dependencies: List[str]):
        """Create task dependencies"""
        for dep_task_id in dependencies:
            dependency = TaskDependency(
                task_id=task_id,
                depends_on_task_id=dep_task_id
            )
            self.db.add(dependency)
        await self.db.commit()
    
    async def schedule_task(self, task_id: str, scheduled_at: datetime) -> bool:
        """Schedule a task for execution"""
        try:
            task = await self.db.get(Task, task_id)
            if not task:
                return False
            
            task.scheduled_at = scheduled_at
            task.status = TaskStatus.SCHEDULED.value
            await self.db.commit()
            return True
        except Exception:
            return False
    
    async def execute_task(self, task_id: str) -> bool:
        """Execute a task immediately"""
        try:
            task = await self.db.get(Task, task_id)
            if not task:
                return False
            
            if task.status not in [TaskStatus.PENDING.value, TaskStatus.SCHEDULED.value]:
                return False
            
            # Check dependencies
            if not await self._check_dependencies(task_id):
                return False
            
            # Start execution
            asyncio.create_task(self._execute_task_async(task))
            return True
        except Exception:
            return False
    
    async def _execute_task_async(self, task: Task):
        """Execute task asynchronously"""
        try:
            # Update task status
            task.status = TaskStatus.RUNNING.value
            task.started_at = datetime.utcnow()
            task.retry_count += 1
            await self.db.commit()
            
            # Create execution log
            execution = TaskExecution(
                task_id=task.id,
                status=TaskStatus.RUNNING.value,
                worker_id=f"worker-{threading.current_thread().ident}"
            )
            self.db.add(execution)
            await self.db.commit()
            
            # Execute task
            handler = self.task_handlers.get(task.type)
            if not handler:
                raise Exception(f"No handler for task type: {task.type}")
            
            start_time = datetime.utcnow()
            result = await handler(task)
            end_time = datetime.utcnow()
            duration = (end_time - start_time).total_seconds()
            
            # Update task with result
            task.status = TaskStatus.COMPLETED.value
            task.completed_at = end_time
            task.result = result
            await self.db.commit()
            
            # Update execution log
            execution.status = TaskStatus.COMPLETED.value
            execution.completed_at = end_time
            execution.duration = duration
            execution.result = result
            await self.db.commit()
            
        except Exception as e:
            # Handle task failure
            await self._handle_task_failure(task, str(e))
    
    async def _handle_task_failure(self, task: Task, error: str):
        """Handle task execution failure"""
        try:
            task.error_message = error
            
            if task.retry_count < task.max_retries:
                # Retry task
                task.status = TaskStatus.RETRYING.value
                retry_delay = task.config.get('retry_delay', 60)
                
                # Schedule retry
                asyncio.create_task(self._retry_task(task, retry_delay))
            else:
                # Mark as failed
                task.status = TaskStatus.FAILED.value
                task.completed_at = datetime.utcnow()
            
            await self.db.commit()
            
        except Exception as e:
            print(f"Error handling task failure: {e}")
    
    async def _retry_task(self, task: Task, delay: int):
        """Retry a failed task after delay"""
        await asyncio.sleep(delay)
        await self._execute_task_async(task)
    
    async def _check_dependencies(self, task_id: str) -> bool:
        """Check if task dependencies are satisfied"""
        dependencies = await self.db.query(TaskDependency).filter(
            TaskDependency.task_id == task_id
        ).all()
        
        for dep in dependencies:
            dep_task = await self.db.get(Task, dep.depends_on_task_id)
            if not dep_task or dep_task.status != TaskStatus.COMPLETED.value:
                return False
        
        return True
    
    # Task Handlers
    async def _handle_email_task(self, task: Task) -> Dict[str, Any]:
        """Handle email task"""
        payload = task.payload
        # Implementation for sending emails
        return {"emails_sent": 1, "recipients": payload.get('recipients', [])}
    
    async def _handle_notification_task(self, task: Task) -> Dict[str, Any]:
        """Handle notification task"""
        payload = task.payload
        # Implementation for sending notifications
        return {"notifications_sent": 1, "channels": payload.get('channels', [])}
    
    async def _handle_report_task(self, task: Task) -> Dict[str, Any]:
        """Handle report generation task"""
        payload = task.payload
        # Implementation for generating reports
        return {"report_generated": True, "file_path": f"/reports/{task.id}.pdf"}
    
    async def _handle_cleanup_task(self, task: Task) -> Dict[str, Any]:
        """Handle cleanup task"""
        payload = task.payload
        # Implementation for cleanup operations
        return {"cleanup_completed": True, "items_cleaned": payload.get('count', 0)}
    
    async def _handle_backup_task(self, task: Task) -> Dict[str, Any]:
        """Handle backup task"""
        payload = task.payload
        # Implementation for backup operations
        return {"backup_completed": True, "backup_path": f"/backups/{task.id}.zip"}
    
    async def _handle_sync_task(self, task: Task) -> Dict[str, Any]:
        """Handle sync task"""
        payload = task.payload
        # Implementation for sync operations
        return {"sync_completed": True, "records_synced": payload.get('count', 0)}
    
    async def _handle_analysis_task(self, task: Task) -> Dict[str, Any]:
        """Handle analysis task"""
        payload = task.payload
        # Implementation for analysis operations
        return {"analysis_completed": True, "insights": payload.get('insights', [])}
    
    async def _handle_custom_task(self, task: Task) -> Dict[str, Any]:
        """Handle custom task"""
        payload = task.payload
        # Implementation for custom tasks
        return {"custom_task_completed": True, "result": payload.get('result', {})}
    
    async def start_scheduler(self):
        """Start the task scheduler"""
        if self.scheduler_running:
            return
        
        self.scheduler_running = True
        asyncio.create_task(self._scheduler_loop())
    
    async def stop_scheduler(self):
        """Stop the task scheduler"""
        self.scheduler_running = False
    
    async def _scheduler_loop(self):
        """Main scheduler loop"""
        while self.scheduler_running:
            try:
                # Get scheduled tasks
                now = datetime.utcnow()
                scheduled_tasks = await self.db.query(Task).filter(
                    Task.status == TaskStatus.SCHEDULED.value,
                    Task.scheduled_at <= now
                ).all()
                
                # Execute scheduled tasks
                for task in scheduled_tasks:
                    await self.execute_task(task.id)
                
                # Wait before next check
                await asyncio.sleep(10)  # Check every 10 seconds
                
            except Exception as e:
                print(f"Scheduler error: {e}")
                await asyncio.sleep(30)  # Wait longer on error
    
    async def get_task_status(self, task_id: str) -> Optional[Dict[str, Any]]:
        """Get task status and details"""
        try:
            task = await self.db.get(Task, task_id)
            if not task:
                return None
            
            return {
                "id": task.id,
                "name": task.name,
                "status": task.status,
                "priority": task.priority,
                "created_at": task.created_at.isoformat(),
                "started_at": task.started_at.isoformat() if task.started_at else None,
                "completed_at": task.completed_at.isoformat() if task.completed_at else None,
                "retry_count": task.retry_count,
                "error_message": task.error_message,
                "result": task.result
            }
        except Exception:
            return None
    
    async def get_tasks(self, status: str = None, task_type: str = None, limit: int = 100) -> List[Dict[str, Any]]:
        """Get tasks with filters"""
        try:
            query = self.db.query(Task)
            
            if status:
                query = query.filter(Task.status == status)
            if task_type:
                query = query.filter(Task.type == task_type)
            
            tasks = await query.order_by(Task.created_at.desc()).limit(limit).all()
            
            return [
                {
                    "id": task.id,
                    "name": task.name,
                    "type": task.type,
                    "status": task.status,
                    "priority": task.priority,
                    "created_at": task.created_at.isoformat(),
                    "scheduled_at": task.scheduled_at.isoformat() if task.scheduled_at else None
                }
                for task in tasks
            ]
        except Exception:
            return []
    
    async def cancel_task(self, task_id: str) -> bool:
        """Cancel a task"""
        try:
            task = await self.db.get(Task, task_id)
            if not task:
                return False
            
            if task.status in [TaskStatus.PENDING.value, TaskStatus.SCHEDULED.value]:
                task.status = TaskStatus.CANCELLED.value
                task.completed_at = datetime.utcnow()
                await self.db.commit()
                return True
            
            return False
        except Exception:
            return False
    
    async def get_task_statistics(self) -> Dict[str, Any]:
        """Get task statistics"""
        try:
            total_tasks = await self.db.query(Task).count()
            pending_tasks = await self.db.query(Task).filter(Task.status == TaskStatus.PENDING.value).count()
            running_tasks = await self.db.query(Task).filter(Task.status == TaskStatus.RUNNING.value).count()
            completed_tasks = await self.db.query(Task).filter(Task.status == TaskStatus.COMPLETED.value).count()
            failed_tasks = await self.db.query(Task).filter(Task.status == TaskStatus.FAILED.value).count()
            
            return {
                "total_tasks": total_tasks,
                "pending_tasks": pending_tasks,
                "running_tasks": running_tasks,
                "completed_tasks": completed_tasks,
                "failed_tasks": failed_tasks,
                "success_rate": (completed_tasks / total_tasks * 100) if total_tasks > 0 else 0
            }
        except Exception:
            return {}
    
    async def cleanup_old_tasks(self, days: int = 30) -> int:
        """Clean up old completed tasks"""
        try:
            cutoff_date = datetime.utcnow() - timedelta(days=days)
            
            old_tasks = await self.db.query(Task).filter(
                Task.status.in_([TaskStatus.COMPLETED.value, TaskStatus.FAILED.value, TaskStatus.CANCELLED.value]),
                Task.completed_at < cutoff_date
            ).all()
            
            count = len(old_tasks)
            for task in old_tasks:
                await self.db.delete(task)
            
            await self.db.commit()
            return count
        except Exception:
            return 0

File: backend/src/task_management.ts
/**
 * TASK MANAGEMENT SYSTEM
 * Advanced task scheduling and execution system for Buffr Host
 */

import { v4 as uuidv4 } from 'uuid';

// Enums
export enum TaskStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  RETRYING = 'retrying'
}

export enum TaskPriority {
  LOW = 'low',
  NORMAL = 'normal',
  HIGH = 'high',
  URGENT = 'urgent',
  CRITICAL = 'critical'
}

export enum TaskType {
  BACKGROUND = 'background',
  SCHEDULED = 'scheduled',
  IMMEDIATE = 'immediate',
  BATCH = 'batch',
  REAL_TIME = 'real_time',
  MAINTENANCE = 'maintenance',
  CLEANUP = 'cleanup',
  SYNC = 'sync',
  NOTIFICATION = 'notification',
  REPORT = 'report',
  CUSTOM = 'custom'
}

// Interfaces
export interface TaskConfig {
  max_retries: number;
  retry_delay: number; // in milliseconds
  timeout: number; // in milliseconds
  priority: TaskPriority;
  parallel_execution: boolean;
  dependencies: string[]; // Task IDs that must complete first
  metadata: Record<string, any>;
}

export interface Task {
  id: string;
  name: string;
  description?: string;
  type: TaskType;
  status: TaskStatus;
  priority: TaskPriority;
  config: TaskConfig;
  payload: Record<string, any>;
  result?: Record<string, any>;
  error_message?: string;
  created_by: string;
  created_at: Date;
  updated_at: Date;
  scheduled_at?: Date;
  started_at?: Date;
  completed_at?: Date;
  retry_count: number;
  max_retries: number;
}

export interface TaskDependency {
  id: string;
  task_id: string;
  depends_on_task_id: string;
  created_at: Date;
}

export interface TaskExecution {
  id: string;
  task_id: string;
  started_at: Date;
  completed_at?: Date;
  status: TaskStatus;
  result?: Record<string, any>;
  error_message?: string;
  execution_time: number; // in milliseconds
  memory_usage?: number; // in MB
  cpu_usage?: number; // percentage
}

export interface TaskManagerOptions {
  db: any; // Database session/connection
  maxWorkers?: number;
  taskTimeout?: number;
  retryDelay?: number;
}

export class TaskManager {
  private db: any;
  private maxWorkers: number;
  private taskTimeout: number;
  private retryDelay: number;
  private workers: Map<string, Worker> = new Map();
  private taskHandlers: Map<string, (payload: Record<string, any>) => Promise<Record<string, any>>> = new Map();
  private isRunning: boolean = false;
  private schedulerInterval?: NodeJS.Timeout;

  constructor(options: TaskManagerOptions) {
    this.db = options.db;
    this.maxWorkers = options.maxWorkers || 5;
    this.taskTimeout = options.taskTimeout || 300000; // 5 minutes default
    this.retryDelay = options.retryDelay || 60000; // 1 minute default
    this.registerDefaultHandlers();
  }

  private registerDefaultHandlers(): void {
    // Task handlers
    this.taskHandlers.set('send_email', this.handleSendEmail.bind(this));
    this.taskHandlers.set('generate_report', this.handleGenerateReport.bind(this));
    this.taskHandlers.set('cleanup_data', this.handleCleanupData.bind(this));
    this.taskHandlers.set('backup_database', this.handleBackupDatabase.bind(this));
    this.taskHandlers.set('sync_integration', this.handleSyncIntegration.bind(this));
    this.taskHandlers.set('process_payment', this.handleProcessPayment.bind(this));
    this.taskHandlers.set('send_notification', this.handleSendNotification.bind(this));
    this.taskHandlers.set('update_analytics', this.handleUpdateAnalytics.bind(this));
    this.taskHandlers.set('custom_task', this.handleCustomTask.bind(this));
  }

  async createTask(
    name: string,
    type: TaskType,
    payload: Record<string, any>,
    config: Partial<TaskConfig> = {},
    createdBy: string = 'system',
    scheduledAt?: Date
  ): Promise<Task> {
    try {
      const taskConfig: TaskConfig = {
        max_retries: 3,
        retry_delay: this.retryDelay,
        timeout: this.taskTimeout,
        priority: TaskPriority.NORMAL,
        parallel_execution: false,
        dependencies: [],
        metadata: {},
        ...config
      };

      const task: Task = {
        id: uuidv4(),
        name,
        description: payload.description || '',
        type,
        status: TaskStatus.PENDING,
        priority: taskConfig.priority,
        config: taskConfig,
        payload,
        created_by: createdBy,
        created_at: new Date(),
        updated_at: new Date(),
        scheduled_at: scheduledAt,
        retry_count: 0,
        max_retries: taskConfig.max_retries
      };

      // Insert into database
      await this.db.query(
        `INSERT INTO tasks (
          id, name, description, type, status, priority, config, payload,
          created_by, created_at, updated_at, scheduled_at, retry_count, max_retries
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
          task.id, task.name, task.description, task.type, task.status, task.priority,
          JSON.stringify(task.config), JSON.stringify(task.payload), task.created_by,
          task.created_at, task.updated_at, task.scheduled_at, task.retry_count, task.max_retries
        ]
      );

      // Create dependencies
      if (taskConfig.dependencies.length > 0) {
        for (const dependencyId of taskConfig.dependencies) {
          await this.createTaskDependency(task.id, dependencyId);
        }
      }

      return task;
    } catch (error) {
      throw new Error(`Failed to create task: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  private async createTaskDependency(taskId: string, dependsOnTaskId: string): Promise<void> {
    try {
      const dependency: TaskDependency = {
        id: uuidv4(),
        task_id: taskId,
        depends_on_task_id: dependsOnTaskId,
        created_at: new Date()
      };

      await this.db.query(
        'INSERT INTO task_dependencies (id, task_id, depends_on_task_id, created_at) VALUES (?, ?, ?, ?)',
        [dependency.id, dependency.task_id, dependency.depends_on_task_id, dependency.created_at]
      );
    } catch (error) {
      console.error('Error creating task dependency:', error);
    }
  }

  async startScheduler(): Promise<void> {
    if (this.isRunning) {
      return;
    }

    this.isRunning = true;
    this.schedulerInterval = setInterval(() => {
      this.schedulerLoop();
    }, 5000); // Check every 5 seconds
  }

  async stopScheduler(): Promise<void> {
    this.isRunning = false;
    if (this.schedulerInterval) {
      clearInterval(this.schedulerInterval);
      this.schedulerInterval = undefined;
    }

    // Stop all workers
    for (const worker of this.workers.values()) {
      worker.stop();
    }
    this.workers.clear();
  }

  private async schedulerLoop(): Promise<void> {
    try {
      // Get pending tasks
      const pendingTasks = await this.getPendingTasks();
      
      // Filter tasks that are ready to run (dependencies satisfied)
      const readyTasks = [];
      for (const task of pendingTasks) {
        if (await this.areDependenciesSatisfied(task.id)) {
          readyTasks.push(task);
        }
      }

      // Sort by priority and creation time
      readyTasks.sort((a, b) => {
        const priorityOrder = { [TaskPriority.CRITICAL]: 5, [TaskPriority.URGENT]: 4, [TaskPriority.HIGH]: 3, [TaskPriority.NORMAL]: 2, [TaskPriority.LOW]: 1 };
        const aPriority = priorityOrder[a.priority] || 0;
        const bPriority = priorityOrder[b.priority] || 0;
        
        if (aPriority !== bPriority) {
          return bPriority - aPriority;
        }
        return a.created_at.getTime() - b.created_at.getTime();
      });

      // Execute tasks up to max workers
      const availableWorkers = this.maxWorkers - this.workers.size;
      const tasksToExecute = readyTasks.slice(0, availableWorkers);

      for (const task of tasksToExecute) {
        this.executeTask(task);
      }
    } catch (error) {
      console.error('Scheduler error:', error);
    }
  }

  private async getPendingTasks(): Promise<Task[]> {
    try {
      const now = new Date();
      const result = await this.db.query(
        'SELECT * FROM tasks WHERE status = ? AND (scheduled_at IS NULL OR scheduled_at <= ?) ORDER BY priority DESC, created_at ASC',
        [TaskStatus.PENDING, now]
      );
      return result.map((row: any) => this.mapRowToTask(row));
    } catch (error) {
      console.error('Error getting pending tasks:', error);
      return [];
    }
  }

  private async areDependenciesSatisfied(taskId: string): Promise<boolean> {
    try {
      const result = await this.db.query(
        'SELECT depends_on_task_id FROM task_dependencies WHERE task_id = ?',
        [taskId]
      );

      if (result.length === 0) {
        return true; // No dependencies
      }

      for (const row of result) {
        const dependencyTask = await this.getTask(row.depends_on_task_id);
        if (!dependencyTask || dependencyTask.status !== TaskStatus.COMPLETED) {
          return false;
        }
      }

      return true;
    } catch (error) {
      console.error('Error checking dependencies:', error);
      return false;
    }
  }

  private async executeTask(task: Task): Promise<void> {
    const worker = new Worker(task, this);
    this.workers.set(task.id, worker);
    
    try {
      await worker.execute();
    } finally {
      this.workers.delete(task.id);
    }
  }

  async getTask(taskId: string): Promise<Task | null> {
    try {
      const result = await this.db.query('SELECT * FROM tasks WHERE id = ?', [taskId]);
      if (result.length === 0) {
        return null;
      }
      return this.mapRowToTask(result[0]);
    } catch (error) {
      console.error('Error getting task:', error);
      return null;
    }
  }

  async updateTaskStatus(
    taskId: string,
    status: TaskStatus,
    result?: Record<string, any>,
    errorMessage?: string
  ): Promise<boolean> {
    try {
      const updateFields: string[] = ['status = ?', 'updated_at = ?'];
      const params: any[] = [status, new Date()];

      if (result) {
        updateFields.push('result = ?');
        params.push(JSON.stringify(result));
      }

      if (errorMessage) {
        updateFields.push('error_message = ?');
        params.push(errorMessage);
      }

      if (status === TaskStatus.RUNNING) {
        updateFields.push('started_at = ?');
        params.push(new Date());
      } else if (status === TaskStatus.COMPLETED || status === TaskStatus.FAILED) {
        updateFields.push('completed_at = ?');
        params.push(new Date());
      }

      params.push(taskId);

      await this.db.query(
        `UPDATE tasks SET ${updateFields.join(', ')} WHERE id = ?`,
        params
      );

      return true;
    } catch (error) {
      console.error('Error updating task status:', error);
      return false;
    }
  }

  async cancelTask(taskId: string, reason?: string): Promise<boolean> {
    try {
      const task = await this.getTask(taskId);
      if (!task || task.status === TaskStatus.COMPLETED || task.status === TaskStatus.FAILED) {
        return false;
      }

      // Stop worker if running
      const worker = this.workers.get(taskId);
      if (worker) {
        worker.stop();
        this.workers.delete(taskId);
      }

      // Update task status
      await this.updateTaskStatus(taskId, TaskStatus.CANCELLED, undefined, reason);

      return true;
    } catch (error) {
      console.error('Error cancelling task:', error);
      return false;
    }
  }

  async retryTask(taskId: string): Promise<boolean> {
    try {
      const task = await this.getTask(taskId);
      if (!task || task.retry_count >= task.max_retries) {
        return false;
      }

      // Update retry count
      await this.db.query(
        'UPDATE tasks SET retry_count = ?, status = ?, updated_at = ? WHERE id = ?',
        [task.retry_count + 1, TaskStatus.RETRYING, new Date(), taskId]
      );

      // Schedule retry
      setTimeout(() => {
        this.db.query(
          'UPDATE tasks SET status = ?, updated_at = ? WHERE id = ?',
          [TaskStatus.PENDING, new Date(), taskId]
        );
      }, task.config.retry_delay);

      return true;
    } catch (error) {
      console.error('Error retrying task:', error);
      return false;
    }
  }

  async getTasks(
    status?: TaskStatus,
    type?: TaskType,
    priority?: TaskPriority,
    limit: number = 100
  ): Promise<Task[]> {
    try {
      let query = 'SELECT * FROM tasks WHERE 1=1';
      const params: any[] = [];

      if (status) {
        query += ' AND status = ?';
        params.push(status);
      }
      if (type) {
        query += ' AND type = ?';
        params.push(type);
      }
      if (priority) {
        query += ' AND priority = ?';
        params.push(priority);
      }

      query += ' ORDER BY created_at DESC LIMIT ?';
      params.push(limit);

      const result = await this.db.query(query, params);
      return result.map((row: any) => this.mapRowToTask(row));
    } catch (error) {
      console.error('Error getting tasks:', error);
      return [];
    }
  }

  async getTaskExecutions(taskId: string, limit: number = 100): Promise<TaskExecution[]> {
    try {
      const result = await this.db.query(
        'SELECT * FROM task_executions WHERE task_id = ? ORDER BY started_at DESC LIMIT ?',
        [taskId, limit]
      );
      return result.map((row: any) => this.mapRowToTaskExecution(row));
    } catch (error) {
      console.error('Error getting task executions:', error);
      return [];
    }
  }

  async getTaskStatistics(): Promise<Record<string, any>> {
    try {
      const totalTasks = (await this.db.query('SELECT COUNT(*) as count FROM tasks'))[0].count;
      const pendingTasks = (await this.db.query('SELECT COUNT(*) as count FROM tasks WHERE status = ?', [TaskStatus.PENDING]))[0].count;
      const runningTasks = (await this.db.query('SELECT COUNT(*) as count FROM tasks WHERE status = ?', [TaskStatus.RUNNING]))[0].count;
      const completedTasks = (await this.db.query('SELECT COUNT(*) as count FROM tasks WHERE status = ?', [TaskStatus.COMPLETED]))[0].count;
      const failedTasks = (await this.db.query('SELECT COUNT(*) as count FROM tasks WHERE status = ?', [TaskStatus.FAILED]))[0].count;

      // Count by type
      const typeCounts: Record<string, number> = {};
      for (const type of Object.values(TaskType)) {
        const count = (await this.db.query('SELECT COUNT(*) as count FROM tasks WHERE type = ?', [type]))[0].count;
        typeCounts[type] = count;
      }

      // Count by priority
      const priorityCounts: Record<string, number> = {};
      for (const priority of Object.values(TaskPriority)) {
        const count = (await this.db.query('SELECT COUNT(*) as count FROM tasks WHERE priority = ?', [priority]))[0].count;
        priorityCounts[priority] = count;
      }

      return {
        total_tasks: totalTasks,
        pending_tasks: pendingTasks,
        running_tasks: runningTasks,
        completed_tasks: completedTasks,
        failed_tasks: failedTasks,
        by_type: typeCounts,
        by_priority: priorityCounts
      };
    } catch (error) {
      console.error('Error getting task statistics:', error);
      return {};
    }
  }

  // Task Handlers
  private async handleSendEmail(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for sending emails
    return { emails_sent: 1, recipients: payload.recipients || [] };
  }

  private async handleGenerateReport(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for generating reports
    return { report_generated: true, file_path: `/reports/${uuidv4()}.pdf` };
  }

  private async handleCleanupData(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for data cleanup
    return { cleanup_completed: true, records_cleaned: payload.count || 0 };
  }

  private async handleBackupDatabase(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for database backup
    return { backup_completed: true, backup_path: `/backups/${uuidv4()}.sql` };
  }

  private async handleSyncIntegration(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for integration sync
    return { sync_completed: true, records_synced: payload.count || 0 };
  }

  private async handleProcessPayment(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for payment processing
    return { payment_processed: true, transaction_id: uuidv4() };
  }

  private async handleSendNotification(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for sending notifications
    return { notifications_sent: 1, channels: payload.channels || [] };
  }

  private async handleUpdateAnalytics(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for analytics update
    return { analytics_updated: true, metrics_processed: payload.count || 0 };
  }

  private async handleCustomTask(payload: Record<string, any>): Promise<Record<string, any>> {
    // Implementation for custom tasks
    return { custom_task_completed: true, result: payload.result || {} };
  }

  // Helper methods
  private mapRowToTask(row: any): Task {
    return {
      id: row.id,
      name: row.name,
      description: row.description,
      type: row.type as TaskType,
      status: row.status as TaskStatus,
      priority: row.priority as TaskPriority,
      config: JSON.parse(row.config || '{}'),
      payload: JSON.parse(row.payload || '{}'),
      result: row.result ? JSON.parse(row.result) : undefined,
      error_message: row.error_message,
      created_by: row.created_by,
      created_at: new Date(row.created_at),
      updated_at: new Date(row.updated_at),
      scheduled_at: row.scheduled_at ? new Date(row.scheduled_at) : undefined,
      started_at: row.started_at ? new Date(row.started_at) : undefined,
      completed_at: row.completed_at ? new Date(row.completed_at) : undefined,
      retry_count: row.retry_count,
      max_retries: row.max_retries
    };
  }

  private mapRowToTaskExecution(row: any): TaskExecution {
    return {
      id: row.id,
      task_id: row.task_id,
      started_at: new Date(row.started_at),
      completed_at: row.completed_at ? new Date(row.completed_at) : undefined,
      status: row.status as TaskStatus,
      result: row.result ? JSON.parse(row.result) : undefined,
      error_message: row.error_message,
      execution_time: row.execution_time,
      memory_usage: row.memory_usage,
      cpu_usage: row.cpu_usage
    };
  }
}

// Worker class for task execution
class Worker {
  private task: Task;
  private taskManager: TaskManager;
  private isStopped: boolean = false;

  constructor(task: Task, taskManager: TaskManager) {
    this.task = task;
    this.taskManager = taskManager;
  }

  async execute(): Promise<void> {
    try {
      // Update task status to running
      await this.taskManager.updateTaskStatus(this.task.id, TaskStatus.RUNNING);

      // Get task handler
      const handler = this.taskManager['taskHandlers'].get(this.task.type);
      if (!handler) {
        throw new Error(`No handler for task type: ${this.task.type}`);
      }

      // Execute task with timeout
      const result = await Promise.race([
        handler(this.task.payload),
        new Promise((_, reject) => 
          setTimeout(() => reject(new Error('Task timeout')), this.task.config.timeout)
        )
      ]) as Record<string, any>;

      // Update task status to completed
      await this.taskManager.updateTaskStatus(this.task.id, TaskStatus.COMPLETED, result);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      
      // Check if we should retry
      if (this.task.retry_count < this.task.max_retries) {
        await this.taskManager.updateTaskStatus(this.task.id, TaskStatus.FAILED, undefined, errorMessage);
        await this.taskManager.retryTask(this.task.id);
      } else {
        await this.taskManager.updateTaskStatus(this.task.id, TaskStatus.FAILED, undefined, errorMessage);
      }
    }
  }

  stop(): void {
    this.isStopped = true;
  }
}

export default TaskManager;